granulate_utils/linux/ns.py

```diff
@@ -67,11 +68,84 @@ def resolve_proc_root_links(proc_root: str, ns_path: str) -> str:
     return path
 
 
+def get_process_nspid(pid: int) -> int:
+    """
+    :raises NoSuchProcess: If the process doesn't or no longer exists
+    """
+    nspid = _get_process_nspid_by_status_file(pid)
+    if nspid is not None:
+        return nspid
+
+    if is_same_ns(pid, NsType.pid.name):
+        # If we're in the same PID namespace, then the outer PID is also the inner pid (NSpid)
+        return pid
+
+    return _get_process_nspid_by_sched_files(pid)
+
+
+def _get_process_nspid_by_status_file(pid: int) -> Optional[int]:
+    try:
+        with open(f"/proc/{pid}/status") as f:
+            for line in f:
+                fields = line.split()
+                if fields[0] == "NSpid:":
+                    return int(fields[-1])
+
+        return None
+    except (FileNotFoundError, ProcessLookupError) as e:
+        raise NoSuchProcess(pid) from e
+
+
+def _get_process_nspid_by_sched_files(pid: int) -> int:
+    # Old kernel (pre 4.1) doesn't have an NSpid field in their /proc/pid/status file
+    # Instead, we can look through all /proc/*/sched files from inside the process' pid namespace, and due to a bug
+    # (fixed in 4.14) the outer PID is exposed, so we can find the target process by comparing the outer PID
+
+    def _find_inner_pid() -> Optional[int]:
+        pattern = re.compile(r"\((\d+), #threads: ")  # Match example: "java (12329, #threads: 11)"
+
+        procfs = Path("/proc")
+        for procfs_child in procfs.iterdir():
+            is_process_dir = procfs_child.is_dir() and procfs_child.name.isdigit()
+            if not is_process_dir:
+                continue
+
+            try:
+                sched_file_path = procfs_child / "sched"
+                with sched_file_path.open("r") as sched_file:
+                    sched_header_line = sched_file.readline()  # The first line contains the outer PID
+            except (FileNotFoundError, ProcessLookupError):
+                # That's OK, processes might disappear before we get the chance to handle them
+                continue
+
+            match = pattern.search(sched_header_line)
+            if match is not None:
+                outer_pid = int(match.group(1))
+                if outer_pid == pid:
+                    return int(procfs_child.name)
+
+        return None
+
+    # We're searching `/proc`, so we only need to set our mount namespace
+    inner_pid = run_in_ns(["mnt"], _find_inner_pid, pid)
+    if inner_pid is not None:
+        return inner_pid
+
+    # If we weren't able to find the process' nspid, he must have been killed while searching
```

Comment:
Eh, this is racy.
1. Another PID may start meanwhile
2. PID might die and remain zombie, and this will return `True`.

What I suggest:
* Move https://github.com/Granulate/gprofiler/blob/8947923cd2a2560a248151e87d602b3460632cd5/gprofiler/utils.py#L483 to `granulate-utils`
* Make a `Process` object before calling `run_in_ns`, then check it afterwards?

Reply:
I don't see a reason to use the `is_process_running()`.

`psutil`'s `is_running()` function is good in this case, I don't care if the process is zombie or not

Reply:
If the `sched` file exists for zombies - sure. If not, we will fail incorrectly, and then you do need to differentiate.

Reply:
Check and comment in code explaining